Adds a bool is_available() method to the backend contract

raziel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53068 Adds a bool is_available() method to the backend contract**

Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` if ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

Commit:
Adds a bool is_available() method to the backend contract

Adds a ```bool is_available()``` method to the backend contract, and use it to implement the following changes in the ```LoweredModule```.
* compile in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* compile at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* execute will only be executed if is_available returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

Overall, we want to ensure the following:

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. Optionally, if that's for some reason not possible, is_available() could return true and the backend will be responsible for either throwing an error on execution or do a no-op. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this means compile() is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates the Glow code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

iseeyuan commented on test/cpp/jit/test_backend_lib.cpp:
```diff
@@ -6,12 +6,17 @@ namespace jit {
 // necessary to test that the JIT backend registration endpoints and
 // code generation are working correctly. It is not intended to
 // produce numerically correct results.
+template <bool isAvailable>
```
What's the purpose of this template argument? To decide if this backend be built or not at compilation time?

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract, and use it to implement the following changes in the ```LoweredModule```.
* compile in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* compile at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* execute will only be executed if is_available returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. Optionally, if that's for some reason not possible, is_available() could return true and the backend will be responsible for either throwing an error on execution or do a no-op. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this means compile() is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -28,6 +28,10 @@ class backend {
         torch::class_<TBackendInterface>(detail::kBackendsNamespace, name)
             .def(torch::init<>())
+            ._def_unboxed(
+                "is_available",
```
It's to bind the method to python. Would be nice to have it tested by Python.

iseeyuan commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -147,6 +158,15 @@ Module codegen_backend_module(
+  loweredModule.define(
+      R"(
+            def __is_available(self):
+                self.__create_backend()
```
`self.__create_backend()` is called in `__setstate__`. Would this creates another backend instance?

raziel replied:
Yes, I need to expose it such that an instance of the backend is available --just as the old "preprocess" code needed to do this.
Actually just as the old code, there were +1 instance of backend created because indeed a call to self.__create_backend() is necessary to make sure self.__backend exists.

In the ideal case self.__create_backend() will handle the lazy instantiation, but cannot figure out a way to check whether self.__backend has been already instantiated given it's created with a instance of a capsule.

Can you expand more on how optional would work?

The other option I can think of is to add a boolean field just to tag whether self.__backend has been instantiated already.

SplitInfinity replied:
> Can you expand more on how optional would work?

Try setting the type of the `__backend` attribute to `Optional[TBackend]` and initializing it to `None` and then checking that it is `None` before creating it in `__create_backend`.

raziel replied:
Implemented with Optional value. Thanks @SplitInfinity 
Tomorrow I'll implement it with a boolean flag to get a sense of which path is cleaner.

raziel replied:
Ok, so reverted the implementation to not use optional value, and instead we just make sure that the backend instance is created once on __setstate__ for the regular save/load flow, and make a separate call to __create_backend at LoweredModule creation time (so we can call __is_available()).

The reason is two fold:
* This is just much cleaner.
* The mobile interpreter does not support get attribute for things that are not tensors (that cause the CI failure --thanks @iseeyuan for helping debug that). This also prevents implementing the other solution of adding a boolean attribute as flag to check for backend initialization.

SplitInfinity replied:
Wait, wasn't whole point to try to create the backend only once? But right now it's still created twice during lowering?

raziel replied:
That’s right. In the case where the LoweredModule is created there’s still 2 backend instances being created, so we can call ins_available().

The issue is that there’s no way to make the optional argument work for mobile; and adding a bool argument to check for initialization has the same issue: from @iseeyuan ”GET_ATTR only works with parameters (tensors) using slot. Since there’s no source info (to save binary size of source parser), it does not work with other types of attributes.”

Having the extra instance of check is_available() is not so bad since is_available() is pretty much a const function.

The arguably more important flow when the LoweredModule is loaded only creates a single backend instance (at __setstate__ time).

raziel replied:
Ok, I implemented a way to only have 1 instance on any case (I validated that's the case).

Please take a look.

iseeyuan reviewed (commented):
Looks great to me in general. One concern is that if it's necessary to call __create_backend in __is_available function in the module. 

Another question is, would it make sense to expose is_available to users? If it does make sense, could we add some use cases (calling is_available()) to unit tests?

SplitInfinity reviewed (commented):
To me, this concept of *availability* is difficult to understand, partly because it seems to have different meanings on the host and target. Why not add `can_compile` and `can_execute` methods instead?

raziel commented on test/jit/test_backends.py:
```diff
@@ -167,6 +168,45 @@ def test_save_load(self):
+
+        # Test exception is thrown.
+        with self.assertRaisesRegex(Exception, r"Backend is not available."):
+            backend_method = self.lowered_module.__getattr__(function_name)
```
my bad, fixed now.

raziel commented on test/jit/test_backends.py:
```diff
@@ -167,6 +168,45 @@ def test_save_load(self):
+        torch.jit.save(self.lowered_module, buffer)
+        buffer.seek(0)
+        with self.assertRaisesRegex(Exception, r"Backend is not available."):
+            imported = torch.jit.load(buffer, map_location=None)
```
Done

raziel commented on test/jit/test_backends.py:
```diff
@@ -167,6 +168,45 @@ def test_save_load(self):
+        input = torch.randn(5)
+
+        # Test exception is thrown.
+        with self.assertRaisesRegex(Exception, r"Backend is not available."):
```
I tried but couldn't raise RuntimeError.

E.g. if I change all the references of Exception for RuntimeError, including the resolver, I get the following errors:

ERROR: test_execution (jit.test_backends.TestBackends)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/razy/Source/mypytorch/pytorch/torch/testing/_internal/common_utils.py", line 426, in wrapper
    fn(*args, **kwargs)
  File "/Users/razy/Source/mypytorch/pytorch/test/jit/test_backends.py", line 444, in test_execution
    self.basic_module_unavailable_test.test_execution()
  File "/Users/razy/Source/mypytorch/pytorch/test/jit/test_backends.py", line 200, in test_execution
    backend_output = backend_method(*(input, input))
  File "/Users/razy/Source/mypytorch/pytorch/torch/testing/_internal/common_utils.py", line 132, in prof_meth_call
    return prof_callable(meth_call, *args, **kwargs)
  File "/Users/razy/Source/mypytorch/pytorch/torch/testing/_internal/common_utils.py", line 126, in prof_callable
    return callable(*args, **kwargs)
torch.jit.Error: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "<string>", line 10, in forward
                  return _0, _1
                else:
                  raise RuntimeError("Backend is not available.")
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE

RuntimeError: RuntimeError: Backend is not available.


======================================================================
ERROR: test_save_load (jit.test_backends.TestBackends)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/razy/Source/mypytorch/pytorch/torch/testing/_internal/common_utils.py", line 426, in wrapper
    fn(*args, **kwargs)
  File "/Users/razy/Source/mypytorch/pytorch/test/jit/test_backends.py", line 451, in test_save_load
    self.basic_module_unavailable_test.test_save_load()
  File "/Users/razy/Source/mypytorch/pytorch/torch/testing/_internal/common_utils.py", line 426, in wrapper
    fn(*args, **kwargs)
  File "/Users/razy/Source/mypytorch/pytorch/test/jit/test_backends.py", line 209, in test_save_load
    imported = torch.jit.load(buffer)
  File "/Users/razy/Source/mypytorch/pytorch/torch/jit/_serialization.py", line 163, in load
    cpp_module = torch._C.import_ir_module_from_buffer(
torch.jit.Error: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File "code/torch/jit.py", line 29, in __setstate__
      self.__handles = _6
    else:
      ops.prim.RaiseException(_4)
      ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return None
  def forward(self: torch.jit.test_backend_unavailableLoweredModule,

Traceback of TorchScript, original code (most recent call last):
  File "<string>", line 9, in __setstate__
                  self.__handles = self.__backend.compile(self.__processed_module, self.__method_compile_spec)
                else:
                  raise RuntimeError("Backend is not available.")
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE

RuntimeError: RuntimeError: Backend is not available.

SplitInfinity replied:
It seems like it is raised, but the class name is being included in the message?

When you use `raise Exception(...)`, what does the error message say?

raziel replied:
It's the same message: raise RuntimeError("Backend is not available.")

The way I changed the resolver was to still use Exception (I couldn't see a sugared value specific to RuntimeError):

    else if (name == "RuntimeError") {
      return std::make_shared<ExceptionValue>(name);
    }

The assertion just changed from Exception to RuntimeError, like this:

        with self.assertRaisesRegex(RuntimeError, r"Backend is not available."):

Do you see an issue? My guess was the resolver, but I don't know how to return a sugared value for RuntimeError.

SplitInfinity replied:
Is the IR different between the two cases? Based on your earlier message it seems that `raise RuntimeError(...)` is being compiled properly.

raziel replied:
I don't see why the IRs would be different, but I'm not sure how to check that.
Overall the code is the same, except for raising and asserting for RuntimeError vs Exception.

kimishpatel replied:
> It's the same message: raise RuntimeError("Backend is not available.")
> 
> The way I changed the resolver was to still use Exception (I couldn't see a sugared value specific to RuntimeError):
> 
> ```
> else if (name == "RuntimeError") {
>   return std::make_shared<ExceptionValue>(name);
> }
> ```
> 
> The assertion just changed from Exception to RuntimeError, like this:
> 
> ```
>     with self.assertRaisesRegex(RuntimeError, r"Backend is not available."):
> ```
> 
> Do you see an issue? My guess was the resolver, but I don't know how to return a sugared value for RuntimeError.

I would think you will have to change torchscript code to do `raise RuntimeError("Backend is not available.")` instead of `raise Exception("Backend is not available.")`?

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract, and use it to implement the following changes in the ```LoweredModule```.
* compile in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* compile at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* execute will only be executed if is_available returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. Optionally, if that's for some reason not possible, is_available() could return true and the backend will be responsible for either throwing an error on execution or do a no-op. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this means compile() is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

raziel commented:
> To me, this concept of _availability_ is difficult to understand, partly because it seems to have different meanings on the host and target. Why not add `can_compile` and `can_execute` methods instead?

Overall it is not meant to have different meanings on host and on target, basically the logic is if is_available is false then we cannot call compile nor execute not pretty much any other method we can add in the future.

I think part of the confusion stems from the naming of "compile" which should have other name, just like preprocess should probably be renamed as well.

But regardless of the names, is_available() covers can_compile() && can_execute()

raziel commented:
> Looks great to me in general. One concern is that if it's necessary to call __create_backend in __is_available function in the module.
> 
> Another question is, would it make sense to expose is_available to users? If it does make sense, could we add some use cases (calling is_available()) to unit tests?

As discussed offline, __create_backend should handle the logic of only creating the instance once. This is an issue since the original implementation where we were already creating more than one instance per LoweredModule.

Also, I don't think we should expose is_available externally (unless there's a good reason). We've discussed its application for implementing fallback logic, but even then this should probably be handled under the covers without the user having to use the function directly.

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract, and use it to implement the following changes in the ```LoweredModule```.
* compile in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* compile at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* execute will only be executed if is_available returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. Optionally, if that's for some reason not possible, is_available() could return true and the backend will be responsible for either throwing an error on execution or do a no-op. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this means compile() is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -140,13 +150,27 @@ Module codegen_backend_module(
+      R"(
+            def __is_available(self):
+                self.__create_backend()
+                backend = self.__backend
```
This copy may look confusing for readers out of TS. Could we add some comment here

raziel replied:
Not used anymore.

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

iseeyuan reviewed (approved):
It looks clean to me, without adding additional check inside the module.

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -184,7 +189,10 @@ Module codegen_backend_module(
                 self.__method_compile_spec = state[0]
                 self.__processed_module = state[1]
-                self.__create_backend()
+                # state[2] indicates wherethercreate_backend
```
```suggestion
```

SplitInfinity reviewed (approved):
Very clever solution to avoid creating two backend instances!

iseeyuan commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -161,8 +189,14 @@ Module codegen_backend_module(
                 self.__processed_module = state[1]
-                self.__create_backend()
-                self.__handles = self.__backend.compile(self.__processed_module, self.__method_compile_spec)
+                # state[2] indicates wherethercreate_backend
```
Rephrase the comments? The two lines on state[2]

SplitInfinity replied:
Yeah I suggested deleting the first line (seems unfinished)

iseeyuan commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -161,8 +189,14 @@ Module codegen_backend_module(
-                self.__create_backend()
-                self.__handles = self.__backend.compile(self.__processed_module, self.__method_compile_spec)
+                # state[2] indicates wherethercreate_backend
+                # state[2] indicates whether to create the backend instance.
```
I think you actually mean this line.

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

kimishpatel reviewed (commented):
Left a few comments. If you can respond, that would be great.

raziel commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -28,6 +28,10 @@ class backend {
+            ._def_unboxed(
+                "is_available",
+                detail::getIsAvailableFunc<TBackendInterface>(),
+                detail::getIsAvailableSchema())
```
I'm not sure if you meant to say "this is BC", but that's right, this is BC because serialized methods have logic serialized that just depends on two methods in the bound object: compile and execute, which didn't change.

raziel commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -264,11 +300,24 @@ Module codegen_backend_module(
-  auto state = at::ivalue::Tuple::create(
-      method_compile_spec, loweredModule.attr("__processed_module"));
-  loweredModule.run_method("__setstate__", state);
+  // If backend is available, call __setstate__ to ensure that the returned
```
The reason why we cannot do this is because we need to create the backend instance in order to make the is_available() call.

kimishpatel replied:
But __setstate__ is creating the backend, right?

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` is ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]

Commit:
Update on "Adds a bool is_available() method to the backend contract"


Adds a ```bool is_available()``` method to the backend contract: it returns ```true``` if ```compile()``` and ```execute()``` can be called; ```false``` otherwise.

It is used to implement the following changes in the ```LoweredModule```:
* ```compile()``` in ```__setstate__``` will run if ```is_available()```, else ```__setstate__``` throws an exception (“Backend not available.”).
* ```compile()``` at ```LoweredModule``` creation will run if ```is_available()```, else a WARNING will be thrown.
* ```execute()``` will only be executed if ```is_available()``` returns true; else throws an exception (“Backend not available.”).

The goal of these changes is to ensure we have a well defined behaviour for the different combinations of backend availability on-host and on-target.

More specifically, backends may have different capabilities to compile and/or execute the Module, depending whether this happens on-host (i.e. where the program is being written) or on-target (where the program is being executed).

First of all, we know that "preprocess" always takes place, and that only happens on-host at creation time. So, we can assume that any compilation is needed/possible on-host then all of it could be pushed here.

Overall, we want to ensure the following:

**On host**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | On module creation, LoweredModule is generated, with a warning  (since compilation and execution can still take place on-target). On module load, throws an exception (since execution is not possible). |
| No | Yes | This configuration should not be possible. This assumes the full compiler is not available, even if some work was done in preprocess the program cannot be finalized for execution. |
| Yes | No | In this case, the expectation would be for is_available() to return false, and compilation logic to move into preprocess. |
| Yes | Yes | All good. This is the only case that is_available() should return true. |

**On target**

| compile | execute | Outcome |
| -- | -- | -- |
| No | No | Loading the LoweredModule throws an exception. Since execution is not possible. |
| No | Yes | Basically this is another instance of Yes/Yes: compilation per se may not be possible on device, which means compile() can be called without issue but it is a no-op, and thus is_available should return true. Consequently, loading the LoweredModule: Succeeds, if the preprocessed module is ready for execution. Fails with exception otherwise. |
| Yes | No | This configuration should not be possible. Just putting here for completeness. |
| Yes | Yes | All good. This, along with No/Yes case (because compilation is assumed to have happened on-host, so it's just another instance of Yes/Yes), are the cases where is_available() should return true. |


**Refactoring existing code**
This change also updates other backends (Glow) code, to implement the is_available() method to have the same behaviour as before this change (i.e. always available).

This should not cause backward incompatibilities with already saved models since we're adding a new method to the PyTorchBackendInterface.
Models saved with the old interface that didn't have is_available() will still find the other 2 methods in the bound object (i.e. compile and execute), and the saved LoweredModule logic will be the old one.

**Future**
We plan to use is_available() to implement support for fallback to the PyTorch interpreter.

Differential Revision: [D26615833](https://our.internmc.facebook.com/intern/diff/D26615833/)

[ghstack-poisoned]
